typescript/core/src/parse.ts

```diff
@@ -1,374 +1,415 @@
-
-  switch (opcode) {
-    case Opcode.HEADER: {
-      const profile = reader.string();
-      const library = reader.string();
-      const record: TypedMcapRecord = { type: "Header", profile, library };
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
+  reader.offset = startOffset + recordLength;
 
-    case Opcode.FOOTER: {
-      const summaryStart = reader.uint64();
-      const summaryOffsetStart = reader.uint64();
-      const summaryCrc = reader.uint32();
-      const record: TypedMcapRecord = {
-        type: "Footer",
-        summaryStart,
-        summaryOffsetStart,
-        summaryCrc,
-      };
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
+  return {
+    type: "Channel",
+    id: channelId,
+    schemaId,
+    topic: topicName,
+    messageEncoding,
+    metadata,
+  };
+}
 
-    case Opcode.SCHEMA: {
-      const id = reader.uint16();
-      const name = reader.string();
-      const encoding = reader.string();
-      const dataLen = reader.uint32();
-      if (reader.offset + dataLen > recordView.byteLength) {
-        throw new Error(`Schema data length ${dataLen} exceeds bounds of record`);
-      }
-      const data = new Uint8Array(
-        recordView.buffer,
-        recordView.byteOffset + reader.offset,
-        dataLen,
-      ).slice();
-      reader.offset += dataLen;
-
-      const record: TypedMcapRecord = {
-        type: "Schema",
-        id,
-        encoding,
-        name,
-        data,
-      };
-
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
+function parseMessage(reader: Reader, recordLength: number): TypedMcapRecord {
+  const channelId = reader.uint16();
+  const sequence = reader.uint32();
+  const logTime = reader.uint64();
+  const publishTime = reader.uint64();
+  const data = reader.u8ArrayCopy(recordLength - 22 /*channelId, sequence, logTime, publishTime*/);
+  return {
+    type: "Message",
+    channelId,
+    sequence,
+    logTime,
+    publishTime,
+    data,
+  };
+}
 
-    case Opcode.CHANNEL: {
-      const channelId = reader.uint16();
-      const schemaId = reader.uint16();
-      const topicName = reader.string();
-      const messageEncoding = reader.string();
-      const metadata = reader.map(
-        (r) => r.string(),
-        (r) => r.string(),
-      );
-
-      const record: TypedMcapRecord = {
-        type: "Channel",
-        id: channelId,
-        schemaId,
-        topic: topicName,
-        messageEncoding,
-        metadata,
-      };
-
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
+function parseChunk(reader: Reader, recordLength: number): TypedMcapRecord {
+  const start = reader.offset;
+  const startTime = reader.uint64();
+  const endTime = reader.uint64();
+  const uncompressedSize = reader.uint64();
+  const uncompressedCrc = reader.uint32();
+  const compression = reader.string();
+  const recordByteLength = Number(reader.uint64());
```

Comment:
drive-by nit: I think this would be better named `recordsByteLength` since it's the length of the `records` field. As is, I was confused by how similar it is to the new parameter `recordLength`